src/field_theory/chevalley_warning.lean

```diff
@@ -0,0 +1,195 @@
+/-
+Copyright (c) 2019 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+
+import data.mv_polynomial
+import field_theory.finite
+
+/-!
+# The Chevalley–Warning theorem
+
+This file contains a proof of the Chevalley–Warning theorem.
+Throughout most of this file, `K` denotes a finite field
+and `q` is notation for the cardinality of `K`.
+
+## Main results
+
+1. Let `f` be a multivariate polynomial in finitely many variables (`X s`, `s : σ`)
+   such that the total degree of `f` is less than `(q-1)` times the cardinality of `σ`.
+   Then the evaluation of `f` on all points of `σ → K` (aka `K^σ`) sums to `0`.
+   (`sum_mv_polynomial_eq_zero`)
+2. The Chevalley–Warning theorem (`char_dvd_card_solutions`).
+   Let `f i` be a finite family of multivariate polynomials
+   in finitely many variables (`X s`, `s : σ`) such that
+   the sum of the total degrees of the `f i` is less than the cardinality of `σ`.
+   Then the number of common solutions of the `f i`
+   is divisible by the characteristic of `K`.
+
+## Notation
+
+- `K` is a finite field
+- `q` is notation for the cardinality of `K`
+- `σ` is the indexing type for the variables of a multivariate polynomial ring over `K`
+
+-/
+
+universes u v
+
+open_locale big_operators
+
+section finite_field
+open mv_polynomial function finset finite_field
+
+variables {K : Type*} {σ : Type*} [fintype K] [field K] [fintype σ]
+local notation `q` := fintype.card K
+
+lemma mv_polynomial.exists_degree_lt_card_sub_one (f : mv_polynomial σ K)
```

Comment:
Why not just wrap the whole file in `namespace mv_polynomial`? (Any particular reason why the two main `theorem`s should not be namespaced?)

Reply:
They felt too specific. Also, the family version doesn't take an `mv_polynomial` as argument, so you can't use dot notation. But I don't have a strong opinion on this.

Reply:
Ah, OK. I was originally thinking that more specific statements should go in more specific namespaces, but on reflection, the more specific a statement is, the less likely there will be name conflicts, so leaving it in the root namespace may just lead to less overall verbosity.